File: Non-UI Files/Harin Sankar/ScanEditor/scan_filter/io/mrd_reader.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterator, Sequence

import numpy as np
from matplotlib import pyplot as plt

from ..core.data_models import MRIFrame, MRIFrameSeries


@dataclass(frozen=True)
class MRDHeader:
    frequency_encoding: int
    phase_encoding: int
    n3d: int
    slices: int
    echoes: int
    experiments: int
    datatype_hex: str
    is_complex: bool
    comments: str
    parameters: str

    @property
    def frame_count(self) -> int:
        return max(1, self.n3d * self.slices * self.echoes * self.experiments)


def read_mrd(file_path: str | Path) -> tuple[np.ndarray, MRDHeader]:
    """Parse an MR Solutions *.mrd file into a k-space array and header metadata."""

    path = Path(file_path)
    if not path.exists():
        raise FileNotFoundError(f"MRD file not found: {file_path}")

    with path.open("rb") as handle:
        header_values = np.fromfile(handle, dtype="<i4", count=4)
        if header_values.size != 4:
            raise IOError("Unexpected end of file while reading MRD header")
        nfe, npe, n3d, nslice = map(int, header_values)

        handle.seek(2, 1)
        dattype_raw = np.fromfile(handle, dtype="<i2", count=1)
        if dattype_raw.size != 1:
            raise IOError("Missing MRD datatype entry")
        dattype_hex = f"{int(dattype_raw[0]):X}".zfill(2)
        is_complex = dattype_hex[0] == "1"

        handle.seek(132, 1)
        dim_values = np.fromfile(handle, dtype="<i4", count=2)
        if dim_values.size != 2:
            raise IOError("MRD echo/experiment counts missing")
        nechoes, nexps = map(int, dim_values)

        current = handle.tell()
        if current < 256:
            handle.seek(256 - current, 1)

        comments = handle.read(256).decode("ascii", errors="ignore").rstrip("\x00")

        num_points = nfe * npe * n3d * nslice * nechoes * nexps
        dtype_code = dattype_hex[1]

        base_dtype: np.dtype
        if dtype_code == "0":
            base_dtype = np.dtype("<u1")
        elif dtype_code == "1":
            base_dtype = np.dtype("<i1")
        elif dtype_code in ("2", "3"):
            base_dtype = np.dtype("<i2")
        elif dtype_code == "4":
            base_dtype = np.dtype("<i4")
        elif dtype_code == "5":
            base_dtype = np.dtype("<f4")
        elif dtype_code == "6":
            base_dtype = np.dtype("<f8")
        else:
            raise ValueError(f"Unsupported MRD datatype nibble: {dtype_code}")

        if num_points == 0:
            data = np.array([], dtype=base_dtype)
        else:
            if not is_complex:
                raw = np.fromfile(handle, dtype=base_dtype, count=num_points)
                if raw.size != num_points:
                    raise IOError("Unexpected end of file while reading MRD data")
                data = raw.reshape((nfe, npe, n3d, nslice, nechoes, nexps), order="F")
            else:
                raw = np.fromfile(handle, dtype=base_dtype, count=num_points * 2)
                if raw.size != num_points * 2:
                    raise IOError("Unexpected end of file while reading complex MRD data")
                real = raw[0::2].astype(np.float64)
                imag = raw[1::2].astype(np.float64)
                complex_data = real + 1j * imag
                data = complex_data.reshape((nfe, npe, n3d, nslice, nechoes, nexps), order="F")

        handle.read(120)  # filename slot (unused)
        parameters = handle.read().decode("ascii", errors="ignore")

    if n3d == 1 and nslice == 1 and nexps == 1:
        data = data[:, :, 0, 0, 0, 0]

    header = MRDHeader(
        frequency_encoding=nfe,
        phase_encoding=npe,
        n3d=n3d,
        slices=nslice,
        echoes=nechoes,
        experiments=nexps,
        datatype_hex=dattype_hex,
        is_complex=is_complex,
        comments=comments,
        parameters=parameters,
    )
    return data, header


def reconstruct_mrd_to_png(
    file_path: str | Path,
    *,
    output_path: str | Path | None = None,
    apply_hamming_window: bool = False,
    zero_pad_shape: tuple[int, int] | None = None,
) -> list[Path]:
    """Reconstruct an MRD file and export frame images as PNG files.

    For single-frame MRD files this writes `<stem>.png` by default.
    For multi-frame MRD files this writes `<stem>_000.png`, `<stem>_001.png`, ...
    """

    path = Path(file_path)
    kspace, _ = read_mrd(path)
    reader = MRDSolutionsReader(
        apply_hamming_window=apply_hamming_window,
        zero_pad_shape=zero_pad_shape,
    )
    frame_arrays = list(reader._iter_frames(kspace))
    if not frame_arrays:
        raise ValueError(f"MRD file {file_path} did not contain any frames")

    if output_path is None:
        if len(frame_arrays) == 1:
            targets = [path.with_suffix(".png")]
        else:
            targets = [path.with_name(f"{path.stem}_{idx:03d}.png") for idx in range(len(frame_arrays))]
    else:
        requested = Path(output_path)
        if len(frame_arrays) == 1:
            if requested.suffix.lower() != ".png":
                requested = requested.with_suffix(".png")
            targets = [requested]
        else:
            if requested.suffix:
                raise ValueError("For multi-frame MRD exports, output_path must be a directory path")
            targets = [requested / f"{path.stem}_{idx:03d}.png" for idx in range(len(frame_arrays))]

    for target in targets:
        target.parent.mkdir(parents=True, exist_ok=True)

    for idx, frame_kspace in enumerate(frame_arrays):
        prepared = reader._prepare_kspace(frame_kspace)
        image = reader._reconstruct_image(prepared)
        scaled_image = reader._scale_to_uint16(image)
        plt.imsave(targets[idx], scaled_image, cmap="gray", vmin=0, vmax=65535)

    return targets


class MRDSolutionsReader:
    """Convert MR Solutions *.mrd files into k-space frame arrays."""

    def __init__(
        self,
        *,
        frame_period_ms: float = 40.0,
        start_time_ms: float = 0.0,
        apply_hamming_window: bool = False,
        zero_pad_shape: tuple[int, int] | None = None,
    ) -> None:
        if frame_period_ms <= 0:
            raise ValueError("frame_period_ms must be positive")
        if zero_pad_shape is not None:
            rows, cols = zero_pad_shape
            if rows <= 0 or cols <= 0:
                raise ValueError("zero_pad_shape must contain positive dimensions")
        self.frame_period_ms = float(frame_period_ms)
        self.start_time_ms = float(start_time_ms)
        self.apply_hamming_window = apply_hamming_window
        self.zero_pad_shape = zero_pad_shape

    def read(self, file_path: str | Path) -> MRIFrameSeries:
        path = Path(file_path)
        kspace, header = read_mrd(path)
        frame_arrays = list(self._iter_frames(kspace))
        if not frame_arrays:
            raise ValueError(f"MRD file {file_path} did not contain any frames")

        base_metadata = {
            "source_path": str(path),
            "source_type": "mrd_kspace",
            "mrd_datatype": header.datatype_hex,
            "mrd_comments": header.comments,
            "mrd_parameters": header.parameters,
            "kspace_shape": list(kspace.shape),
        }

        frames: list[MRIFrame] = []
        for idx, frame_kspace in enumerate(frame_arrays):
            prepared = self._prepare_kspace(frame_kspace)
            timestamp_ms = self.start_time_ms + idx * self.frame_period_ms
            metadata = {
                **base_metadata,
                "frame_index": idx,
                "rows": int(prepared.shape[0]),
                "columns": int(prepared.shape[1]),
            }
            frames.append(
                MRIFrame(
                    frame_id=idx,
                    timestamp_ms=timestamp_ms,
                    dataset=prepared,
                    metadata=metadata,
                ),
            )
        return MRIFrameSeries(frames)

    def _prepare_kspace(self, frame_kspace: np.ndarray) -> np.ndarray:
        data = np.array(frame_kspace, copy=True)
        if self.apply_hamming_window:
            data = self._apply_hamming_window(data)
        if self.zero_pad_shape is not None:
            data = self._zero_pad(data, self.zero_pad_shape)
        return data

    @staticmethod
    def _iter_frames(kspace: np.ndarray) -> Iterator[np.ndarray]:
        if kspace.ndim == 2:
            yield kspace
            return
        fe, pe = kspace.shape[:2]
        trailing = int(np.prod(kspace.shape[2:]))
        reshaped = np.reshape(kspace, (fe, pe, trailing), order="F")
        for idx in range(trailing):
            yield reshaped[:, :, idx]

    @staticmethod
    def _apply_hamming_window(data: np.ndarray) -> np.ndarray:
        rows, cols = data.shape
        wx = np.hamming(rows)
        wy = np.hamming(cols)
        return data * np.outer(wx, wy)

    @staticmethod
    def _zero_pad(data: np.ndarray, target_shape: Sequence[int]) -> np.ndarray:
        rows, cols = data.shape
        target_rows, target_cols = int(target_shape[0]), int(target_shape[1])
        if target_rows < rows or target_cols < cols:
            raise ValueError("zero_pad_shape must be larger than the input dimensions")
        row_pad_before = (target_rows - rows) // 2
        row_pad_after = target_rows - rows - row_pad_before
        col_pad_before = (target_cols - cols) // 2
        col_pad_after = target_cols - cols - col_pad_before
        return np.pad(
            data,
            ((row_pad_before, row_pad_after), (col_pad_before, col_pad_after)),
            mode="constant",
            constant_values=0,
        )

    @staticmethod
    def _reconstruct_image(kspace: np.ndarray) -> np.ndarray:
        return np.abs(np.fft.fftshift(np.fft.fft2(kspace)))

    @staticmethod
    def _scale_to_uint16(image: np.ndarray) -> np.ndarray:
        magnitude = np.abs(image)
        finite_mask = np.isfinite(magnitude)
        if not np.any(finite_mask):
            return np.zeros_like(magnitude, dtype=np.uint16)
        filtered = np.zeros_like(magnitude, dtype=float)
        filtered[finite_mask] = magnitude[finite_mask]
        max_val = float(filtered.max())
        if max_val <= 0:
            return np.zeros_like(filtered, dtype=np.uint16)
        normalized = filtered / max_val
        scaled = np.clip(normalized * 65535.0, 0, 65535)
        return scaled.astype(np.uint16)

File: Non-UI Files/Harin Sankar/ScanEditor/CardiacGatingCS/MrdReader.cs
using System.Numerics; // Complex numbers for k-space.
using System.Text; // ASCII decoding for header strings.

namespace CardiacGating; // Project namespace.

// Immutable MRD header snapshot used by higher-level readers.
public sealed record MrdHeader(
    int FrequencyEncoding, // Number of FE points.
    int PhaseEncoding,     // Number of PE points.
    int N3d,               // Third spatial dimension size.
    int Slices,            // Slice count.
    int Echoes,            // Echo count.
    int Experiments,       // Experiment count.
    string DatatypeHex,    // Raw datatype nibble pair as hex.
    bool IsComplex,        // True when payload stores real/imag pairs.
    string Comments,       // Fixed-size comments block.
    string Parameters)     // Trailing parameter text.
{
    public int FrameCount => Math.Max(1, N3d * Slices * Echoes * Experiments); // Non-spatial frame count.
}

/// <summary>
/// Reads MR Solutions MRD binary files and extracts k-space frames.
/// </summary>
public static class MrdFileReader
{
    // Low-level file parse returning raw k-space frames + parsed header.
    public static (Complex[][,] Frames, MrdHeader Header) Read(string filePath)
    {
        if (!File.Exists(filePath)) // Validate input path before opening stream.
            throw new FileNotFoundException($"MRD file not found: {filePath}");

        using var stream = new FileStream( // Open file stream for binary reading.
            filePath,
            FileMode.Open,
            FileAccess.Read,
            FileShare.Read,
            65536); // Larger buffer for throughput.

        using var reader = new BinaryReader( // Binary primitive reader.
            stream,
            Encoding.ASCII,
            leaveOpen: false);

        int nfe = reader.ReadInt32();    // FE dimension.
        int npe = reader.ReadInt32();    // PE dimension.
        int n3d = reader.ReadInt32();    // 3D dimension.
        int nslice = reader.ReadInt32(); // Slice dimension.

        stream.Seek(2, SeekOrigin.Current); // Skip reserved bytes.

        short dtRaw = reader.ReadInt16();                      // Raw datatype word.
        string dtHex = ((ushort)dtRaw).ToString("X").PadLeft(2, '0'); // Two-nibble datatype representation.
        bool isComplex = dtHex[0] == '1';                      // Leading nibble encodes complex payload flag.
        char dtCode = dtHex[1];                                // Trailing nibble encodes primitive type.

        stream.Seek(132, SeekOrigin.Current); // Skip reserved block.

        int nechoes = reader.ReadInt32(); // Echo dimension.
        int nexps = reader.ReadInt32();   // Experiment dimension.

        stream.Seek(256, SeekOrigin.Begin); // Jump to comments block start.
        byte[] commBytes = reader.ReadBytes(256); // Read fixed comments block.
        string comments = Encoding.ASCII.GetString(commBytes).TrimEnd('\0').Trim(); // Remove null padding.

        long numPoints = (long)nfe * npe * n3d * nslice * nechoes * nexps; // Scalar payload count.
        Complex[] flat = ReadFlatData(reader, dtCode, numPoints, isComplex); // Decode flat payload.

        long filenameEnd = stream.Position + 120; // Optional filename slot end position.
        if (filenameEnd <= stream.Length)         // Only skip if slot exists.
            stream.Seek(120, SeekOrigin.Current);

        long remaining = stream.Length - stream.Position; // Remaining tail bytes.
        string parameters = remaining > 0
            ? Encoding.ASCII.GetString(reader.ReadBytes((int)Math.Min(remaining, int.MaxValue))).TrimEnd('\0') // Tail text.
            : string.Empty; // Empty when no tail remains.

        var header = new MrdHeader( // Build immutable header record.
            nfe,
            npe,
            n3d,
            nslice,
            nechoes,
            nexps,
            dtHex,
            isComplex,
            comments,
            parameters);

        int safe3d = Math.Max(1, n3d);           // Prevent zero in product/decomposition.
        int safeSlice = Math.Max(1, nslice);      // Prevent zero in product/decomposition.
        int safeEchoes = Math.Max(1, nechoes);    // Prevent zero in product/decomposition.
        int safeExps = Math.Max(1, nexps);        // Prevent zero in product/decomposition.
        int trailing = safe3d * safeSlice * safeEchoes * safeExps; // Number of output frames.

        var frames = new Complex[trailing][,]; // Allocate output frame array.

        for (int t = 0; t < trailing; t++) // Convert each trailing-index slice into one 2D frame.
        {
            int tmp = t;                                // Working index for decomposition.
            int i3d = tmp % safe3d; tmp /= safe3d;     // Decode 3D index (fastest varying).
            int islice = tmp % safeSlice; tmp /= safeSlice; // Decode slice index.
            int iecho = tmp % safeEchoes; tmp /= safeEchoes; // Decode echo index.
            int iexp = tmp;                             // Remaining quotient is experiment index.

            var frame = new Complex[nfe, npe]; // Allocate FE x PE matrix.

            for (int ife = 0; ife < nfe; ife++) // FE loop.
            for (int ipe = 0; ipe < npe; ipe++) // PE loop.
            {
                long idx = ife
                         + (long)nfe * (ipe
                         + (long)npe * (i3d
                         + (long)n3d * (islice
                         + (long)nslice * (iecho
                         + (long)nechoes * iexp)))); // Flattened index in MRD column-major layout.

                frame[ife, ipe] = flat[idx]; // Copy scalar into 2D matrix position.
            }

            frames[t] = frame; // Store extracted frame.
        }

        return (frames, header); // Return all frames + header metadata.
    }

    // Decodes payload values and converts into complex array.
    private static Complex[] ReadFlatData(BinaryReader reader, char dtCode, long numPoints, bool isComplex)
    {
        if (numPoints == 0) // Empty payload case.
            return Array.Empty<Complex>();

        long readCount = isComplex ? numPoints * 2 : numPoints; // Complex payload uses real+imag scalars.
        double[] raw = ReadDoubles(reader, dtCode, (int)readCount); // Decode primitive values as doubles.

        var result = new Complex[numPoints]; // Allocate complex output array.

        if (isComplex) // Interpret as interleaved pairs.
        {
            for (long i = 0; i < numPoints; i++)
                result[i] = new Complex(raw[i * 2], raw[i * 2 + 1]); // [real, imag].
        }
        else // Interpret as real-only values.
        {
            for (long i = 0; i < numPoints; i++)
                result[i] = new Complex(raw[i], 0.0); // Imaginary component is zero.
        }

        return result; // Return decoded complex payload.
    }

    // Dispatches datatype code to concrete primitive reader.
    private static double[] ReadDoubles(BinaryReader r, char code, int count)
    {
        return code switch
        {
            '0' => r.ReadBytes(count).Select(b => (double)b).ToArray(), // Unsigned byte.
            '1' => ReadSBytes(r, count),                                 // Signed byte.
            '2' or '3' => ReadInt16s(r, count),                          // Int16 variants.
            '4' => ReadInt32s(r, count),                                 // Int32.
            '5' => ReadFloat32s(r, count),                               // Float32.
            '6' => ReadFloat64s(r, count),                               // Float64.
            _ => throw new InvalidDataException($"Unsupported MRD datatype nibble: {code}") // Unknown code.
        };
    }

    // Reads signed 8-bit values and promotes to double.
    private static double[] ReadSBytes(BinaryReader r, int n)
        => r.ReadBytes(n).Select(b => (double)(sbyte)b).ToArray();

    // Reads int16 values and promotes to double.
    private static double[] ReadInt16s(BinaryReader r, int n)
    {
        var a = new double[n]; // Output array.
        for (int i = 0; i < n; i++)
            a[i] = r.ReadInt16(); // Promote each int16.
        return a;
    }

    // Reads int32 values and promotes to double.
    private static double[] ReadInt32s(BinaryReader r, int n)
    {
        var a = new double[n]; // Output array.
        for (int i = 0; i < n; i++)
            a[i] = r.ReadInt32(); // Promote each int32.
        return a;
    }

    // Reads float32 values and promotes to double.
    private static double[] ReadFloat32s(BinaryReader r, int n)
    {
        var a = new double[n]; // Output array.
        for (int i = 0; i < n; i++)
            a[i] = r.ReadSingle(); // Promote each float32.
        return a;
    }

    // Reads float64 values directly.
    private static double[] ReadFloat64s(BinaryReader r, int n)
    {
        var a = new double[n]; // Output array.
        for (int i = 0; i < n; i++)
            a[i] = r.ReadDouble(); // Copy each float64.
        return a;
    }
}

/// <summary>
/// High-level MRD reader that timestamps frames and applies optional preprocessing.
/// </summary>
public sealed class MrdSolutionsReader
{
    public double FramePeriodMs { get; } // Timestamp delta between consecutive frames.
    public double StartTimeMs { get; } // Timestamp assigned to first frame.
    public bool ApplyHammingWindow { get; } // Whether to apply Hamming apodization.
    public (int Fe, int Pe)? ZeroPadShape { get; } // Optional zero-pad shape.

    // Constructor stores preprocessing/timestamp parameters.
    public MrdSolutionsReader(
        double framePeriodMs = 40.0,
        double startTimeMs = 0.0,
        bool applyHammingWindow = false,
        (int, int)? zeroPadShape = null)
    {
        if (framePeriodMs <= 0) // Frame period must be positive.
            throw new ArgumentOutOfRangeException(nameof(framePeriodMs), "Must be positive");

        FramePeriodMs = framePeriodMs; // Save frame delta.
        StartTimeMs = startTimeMs; // Save first timestamp.
        ApplyHammingWindow = applyHammingWindow; // Save apodization flag.
        ZeroPadShape = zeroPadShape.HasValue ? ((int Fe, int Pe)?)zeroPadShape.Value : null; // Normalize nullable tuple.
    }

    // Reads MRD file, prepares k-space frames, and returns timestamped series.
    public MRIFrameSeries Read(string filePath)
    {
        (Complex[][,] rawFrames, MrdHeader header) = MrdFileReader.Read(filePath); // Parse raw MRD file.

        if (rawFrames.Length == 0) // Empty frame payload is invalid for pipeline use.
            throw new InvalidDataException($"MRD file '{filePath}' contains no frames");

        var baseMeta = new Dictionary<string, object> // Shared metadata copied to each frame.
        {
            ["source_path"] = filePath,
            ["source_type"] = "mrd_kspace",
            ["mrd_datatype"] = header.DatatypeHex,
            ["mrd_comments"] = header.Comments,
            ["mrd_parameters"] = header.Parameters,
        };

        var frames = new List<MRIFrame>(rawFrames.Length); // Output frame list.

        for (int i = 0; i < rawFrames.Length; i++) // Convert each raw frame into MRIFrame object.
        {
            Complex[,] prepared = PrepareKspace(rawFrames[i]); // Apply optional preprocessing.

            var meta = new Dictionary<string, object>(baseMeta) // Copy shared metadata.
            {
                ["frame_index"] = i,
                ["rows"] = prepared.GetLength(0),
                ["columns"] = prepared.GetLength(1),
            };

            frames.Add(new MRIFrame(
                frameId: i,
                timestampMs: StartTimeMs + i * FramePeriodMs, // Deterministic synthetic timestamp.
                dataset: prepared,
                metadata: meta));
        }

        return new MRIFrameSeries(frames); // Wrap in immutable series object.
    }

    // Applies configured preprocessing pipeline to one frame.
    private Complex[,] PrepareKspace(Complex[,] frame)
    {
        Complex[,] data = (Complex[,])frame.Clone(); // Clone so source data is not modified.

        if (ApplyHammingWindow) // Optional apodization stage.
            data = ApplyHamming(data);

        if (ZeroPadShape.HasValue) // Optional zero-padding stage.
            data = ZeroPad(data, ZeroPadShape.Value.Fe, ZeroPadShape.Value.Pe);

        return data; // Return prepared frame matrix.
    }

    // Applies separable 2D Hamming window.
    private static Complex[,] ApplyHamming(Complex[,] data)
    {
        int rows = data.GetLength(0); // Source row count.
        int cols = data.GetLength(1); // Source column count.

        double[] wr = HammingWindow(rows); // Row coefficients.
        double[] wc = HammingWindow(cols); // Column coefficients.

        var result = new Complex[rows, cols]; // Output matrix.

        for (int r = 0; r < rows; r++)
        for (int c = 0; c < cols; c++)
            result[r, c] = data[r, c] * wr[r] * wc[c]; // Apply outer-product coefficient.

        return result; // Return apodized matrix.
    }

    // Generates classic Hamming coefficients for length n.
    private static double[] HammingWindow(int n)
    {
        var w = new double[n]; // Output coefficient vector.

        for (int i = 0; i < n; i++)
            w[i] = 0.54 - 0.46 * Math.Cos(2.0 * Math.PI * i / (n - 1)); // Hamming formula.

        return w; // Return coefficient vector.
    }

    // Zero-pads matrix to new dimensions while centering original content.
    private static Complex[,] ZeroPad(Complex[,] data, int newRows, int newCols)
    {
        int rows = data.GetLength(0); // Source rows.
        int cols = data.GetLength(1); // Source cols.

        if (newRows < rows || newCols < cols) // Zero-pad cannot shrink dimensions.
            throw new ArgumentException("Zero-pad target must be >= input dimensions");

        int pr = (newRows - rows) / 2; // Row insertion offset.
        int pc = (newCols - cols) / 2; // Column insertion offset.

        var result = new Complex[newRows, newCols]; // Padded matrix initialized to zeros.

        for (int r = 0; r < rows; r++)
        for (int c = 0; c < cols; c++)
            result[r + pr, c + pc] = data[r, c]; // Copy source into centered region.

        return result; // Return padded matrix.
    }
}
